megamek/src/megamek/client/bot/princess/Princess.java

```diff
@@ -2157,6 +2170,172 @@ private void launchFighters(MovePath path) {
+                }
+
+                for (WeaponMounted curAMS : activeAMS) {
+
+                    EquipmentMode curMode = curAMS.curMode();
+
+                    // Turn off laser AMS to help with overheating problems
+                    if (curAMS.getType().hasFlag(WeaponType.F_ENERGY)) {
+                        if (isOverheating) {
+                            newAMSMode = EquipmentMode.getMode(Weapon.MODE_AMS_OFF);
+                        }
+                    } else {
+
+                        // Determine if ammo needs to be conserved
+                        boolean conserveAmmo = curAMS.getLinkedAmmo().getUsableShotsLeft() <= (int) Math.floor(curAMS.getOriginalShots() *
+                                behaviorSettings.getSelfPreservationValue() / 100.0);
+
+                        // Consider turning off AMS to conserve ammo unless it's needed for infantry
+                        if (conserveAmmo && !newAMSMode.equals(Weapon.MODE_AMS_MANUAL)) {
+
+                            int ammoTN = 12 - behaviorSettings.getBraveryIndex();
+
+                            // Fighting a missile boat is more likely to require an active AMS
+                            int lastTargetID = curEntity.getLastTarget();
+                            if (lastTargetID >= 0) {
+                                Entity lastTarget = game.getEntity(lastTargetID);
+                                if (lastTarget != null && lastTarget.getRole() == UnitRole.MISSILE_BOAT) {
+                                    ammoTN += 4;
+                                }
+                            }
+
+                            // Heavily damaged units are more likely to require an active AMS than
+                            // lightly damaged ones
+                            switch (curEntity.getDamageLevel()) {
+                                case Entity.DMG_NONE:
+                                    ammoTN -= 4;
+                                    break;
+                                case Entity.DMG_LIGHT:
+                                    ammoTN -= 2;
+                                    break;
+                                case Entity.DMG_MODERATE:
+                                    ammoTN += 1;
+                                    break;
+                                case Entity.DMG_HEAVY:
+                                    ammoTN += 4;
+                                case Entity.DMG_CRIPPLED:
+                                    ammoTN += 8;
+                                    break;
+                                default:
+                                    break;
+                            }
+
+                            if (ammoTN < 10) {
+                                if (Compute.d6(2) >= ammoTN) {
+                                    newAMSMode = EquipmentMode.getMode(Weapon.MODE_AMS_OFF);
+                                }
+                            }
```

Comment:
Why is the choice to change mode randomized rather than deterministic?

Reply:
While it's good to have the bulk of the behavior calculated rather than random, I'm not entirely sold on a fully deterministic Princess; for one, it is too easy to predict.  There will be circumstances like this, where there is no clear 'good' outcome that can be easily calculated vs the impact the decision will have.  Sometimes it's best all 'round to have it to do a simple gut check and hope for the best.